src/interface/state_descriptor.hpp

```diff
@@ -83,6 +70,41 @@ struct VarIDHasher {
   }
 };
 
+/// A little container class owning refinement function properties
+/// needed for the state descriptor.
+/// Note using VarID here implies that custom prolongation/restriction
+/// is identical for all sparse vars in a pool.
+/// Note ignores sparse id, so all sparse ids of a
+/// given sparse name have the same prolongation/restriction
+/// operations
+/// TODO(JMM): Should this cache be a static member field of
+/// RefinementFunctions_t? That would mean we could avoid
+/// StateDescriptor entirely.
+struct RefinementFunctionMaps {
+  void Register(const Metadata &m) {
+    if (m.IsRefined()) {
+      const auto &funcs = m.GetRefinementFunctions();
+      bool in_map = (funcs_to_ids.count(funcs) > 0);
+      if (!in_map) {
+        funcs_to_ids[funcs] = next_refinement_id_++;
+      }
+    }
+  }
+
+  std::size_t size() const noexcept { return next_refinement_id_; }
+  // A unique enumeration of refinement functions starting from zero.
+  // This is used for caching which prolongation/restriction operator
+  // matches which BndInfo struct in the buffer packing caches.
+  // the other relevant information is in metadata, so this is all we
+  // need.
+  std::unordered_map<refinement::RefinementFunctions_t, std::size_t,
+                     refinement::RefinementFunctionsHasher>
+      funcs_to_ids;
```

Comment:
Does this imply that a registered refinement operation may show up more than once in the `funcs_to_ids` list (since it is the prologation/restriction pair that is used for the hash)?

Reply:
Yes---the same restriction op might show up multiple times if it's coupled with a different prolongation function. We could resolve that by splitting prolongation/restriction pairs into separate objects. But I'm not sure how much sense that makes physically. IMO your prolongation and restriction should both be coupled to some assumption about discretization. Maybe I'm wrong about that though.

I'm also not sure it matters too much if that ends up happening anyway.